### DISCUSSION TITLE
How to ask for help with code

### DISCUSSION THREAD

**Discussion by BekahHW** [Virtual-Coffee/virtualcoffee.io#409] [New Ideas]
I'd like to see a section on the member resources section that gives tips on how to ask for help. I thought we could brainstorm here before turning it into an issue. 

I'm thinking a couple of sections:

**Before asking for help**
- Google the error

**When asking for help**
- List what you've tried

What can we add?

**Comment by davidalpert** [comment]
wherever possible, provide replication details, such as:
- what language version or framework version or libraries version are you using when you encounter the error?
- what operating system (e.g. windows, mac, linux) and version are you experiencing the problem on? 
- what command or steps did you run? (i.e. anything I need in order to be able to interpret the error or reproduce it)
- what have you already tried? (it's okay if you don't know where to start but share whatever you have already tried/ruled out)

none of this is intended to make it difficult to ask a question.

a simple "can someone help me with ___" is enough to get a conversation started in a supportive community.

with code problems, however, we often need a lot of the details such as versions or replication steps in order to understand what is happening and how to be helpful

**Comment by jonathanyeong** [comment]
**when asking for help**
- Be specific about what you're looking for. "Does anyone know why my Mongo DB authentication failed" is better than "My app is crashing".
- If you're struggling to articulate the question, ask to get someone to teach you a concept. This might not be specific to your current problem, but it might uncover a thread to pull! In this case, mention what you've tried or found. E.g. "I see this error that's happening with mongo. But I know nothing about it, can someone teach me about mongo?"

**Comment by danieltott** [comment]
I'm not sure if this is the right place for it, but should we add some Virtual Coffee specific language? Like "try the discussion board" or "use help-and-pairing, be sure to thread things, etc etc"?

**Comment by abbeyperini** [comment]
If it's a specific error or something not working, code snippets (in slack or an external link) or screenshots or willingness to jump in the co-working room to talk it through are paramount.

If it's a more general/trying to grasp a concept question, "Does anyone have experience with ___?" or "Can someone explain ___ to me like I'm five?" is a good starting point. Sometimes I just need someone to ask me questions about what I'm looking for to formulate my question/clarify the piece of the puzzle I'm missing/move forward with my own Google search.

**Comment by marktnoonan** [comment]
Often when we need help we've worked way too long on something and are very frustrated by it, so I would say something like this up top:

> Everybody runs into problems during development that can be frustrating and seem impossible to debug. The most important thing is: if you need help, ask for help. Here's a guide to asking for help effectively as a member of Virtual Coffee. If you're still not sure how to ask for help: ask anyway. We've all been there. 

Sometimes what you need is just extra eyes and somebody to think through the problem with you. So in line with the suggestions here, it would be good to keep a focus on the welcoming aspect of VC. So I'd suggest whatever the guidelines are, the tone can be "here are some effective ways to get help depending on what you need", & we can refer people to it, but not "here's the way you must ask for help at VC".

**Comment by abbeyperini** [comment]
Expanding on Mark's excellent points -> maybe something about suggesting a break? We all get tunnel vision sometimes.

**Comment by raykotab** [comment]
I think this thread is very insightful on how VC works. I myself struggle sometimes to find a starting point for asking or communicating. But anyways it ist a lot what I learn and if I don't get more help it is because I lack the time to keep reading messages and being around. Thanks you all

**Comment by LincolnFleet** [comment]
I'm not sure how structured you want to make this resource, but we could make a form with prompts and inputs for each of the info sections we recommend including. When the form is submitted it just renders a copy-to-clipboard-able, plaintext version of the sections and input that have been filled out in a format that's conducive to being posted in Slack or Stackoverflow or wherever. It could serve not just as advice for how to best form a question, but also a framework for posting it in a respectable format.
We could even integrate spellcheck for it! Ok maybe I'm going too far...

**Reply by BekahHW** [reply]
ohhhh. That sounds pretty cool. I like the idea of having the resources and then at the bottom, something like "Try out our form to help you create your coding question." And then we can sell it to Stackoverflow

**Comment by Claire** [comment]
I always like to see what the asker tried already, it shows what their understanding is and gaps, also that they made some effort. Stack overflow has some god best practices [https://stackoverflow.com/help/how-to-ask](https://stackoverflow.com/help/how-to-ask) and a checklist is always easier to consume [https://codeblog.jonskeet.uk/2012/11/24/stack-overflow-question-checklist/](https://codeblog.jonskeet.uk/2012/11/24/stack-overflow-question-checklist/)
